### PR TITLE
Improve security definer for procedures/functions by avoid Lookup for GUC values

### PR DESCRIPTION
**Pull request by shah-nirmit**
### Description

Inside function `fmgr_security_definer` function there were frequent calls to `set_config_option/GetConfigOption` API's for `sql_dialect` GUC which would need Cache Lookup, which is unnecessary for the dialect. Changed the implementation to directly set/get the GUC value using the global variable.

Draft PR on Extension Repo: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/977
### Issues Resolved

BABEL-2994
 
### Test Scenarios Covered ###
* **Use case based -**
Already Existing Test Cases

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**
There would be a general performance for procedures,functions which use the FMGR framework
Here is one specific scenario
 A test database populated with 15k dummy procedures and executed the following query
`exec sp_sproc_columns  @procedure_name='sp_statistics' , @ODBCVer=3, @fUsePattern=0`
We had the following results
```
Without the change
Network packet size (bytes): 4096
50 xact[s]:
Clock Time (ms.): total     16516  avg   330.3 (3.0 xacts per sec.)

vs 

With the change
Network packet size (bytes): 4096
50 xact[s]:
Clock Time (ms.): total      9335  avg   186.7 (5.4 xacts per sec.)
```
Explanation:-For sp_sproc_columns, we he have created an internal view sp_sproc_columns_view for all procedures. The view itself has a lot of cast functions; mostly for each column  and each row for example if we cast to string types then varchar function is called using FMGR framework.
Later we select only one row for particular input based on procedure_name, but even there there’s a mismatch in datatype length so varchar function is called for one column, each row.

Attached the perf reports 
[perf-without](https://user-images.githubusercontent.com/44708230/210231957-03ceaee9-f0f5-413d-af37-27786f46b8f0.svg)
[perf-with-2994](https://user-images.githubusercontent.com/44708230/210231963-4c6bd6cf-5dd0-4726-a0d1-6516e285a086.svg)

* **Tooling impact -**

* **Client tests -**

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/misc/guc.c [hunk 1]
```diff
@@ -4990,6 +4990,9 @@ static struct config_enum ConfigureNamesEnum[] =
 		NULL, NULL, NULL
 	},
 
+/* When changing this guc value directly without set_config_option,The function
```
Comment (surendravishnoi) [inline src/backend/utils/misc/guc.c]: please use
/*
   comment
*/

File: src/backend/utils/fmgr/fmgr.c [hunk 2]
```diff
@@ -821,14 +823,10 @@ fmgr_security_definer(PG_FUNCTION_ARGS)
...
 		ReleaseSysCache(tuple);
-		sql_dialect_value_old = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
-		set_config_option("babelfishpg_tsql.sql_dialect", sql_dialect_value,
-						  (superuser() ? PGC_SUSET : PGC_USERSET),
-						  PGC_S_SESSION,
-						  GUC_ACTION_SAVE,
-						  true,
-						  0,
-						  false);
+		sql_dialect_value_old = sql_dialect;
```
Comment (surendravishnoi) [inline src/backend/utils/fmgr/fmgr.c:826]: do we have some simple test case which shows benefit of this change? Ideally, it should be a function/proc where cross dialect calls are done in each statement and the statements are executed in a loop

**Review by surendravishnoi — COMMENTED** [review]
Use PR template

File: src/backend/utils/misc/guc.c (hunk 1 above)
Comment (shah-nirmit) [reply src/backend/utils/misc/guc.c]: Fixed in the latest commit

**Comment by KushaalShroff** [comment]
> Use PR template

@shah-nirmit Please add one or more of the scenarios where this was a bottleneck and the perfomance impact % numbers for them

**Review by surendravishnoi — DISMISSED** [review]

File: src/backend/utils/fmgr/fmgr.c (hunk 2 above)
Comment (shah-nirmit) [reply src/backend/utils/fmgr/fmgr.c:826]: Addressed in PR description

**Review by KushaalShroff — APPROVED** [review]
